viz/TraversabilityGridVisualization.cpp

```diff
@@ -0,0 +1,321 @@
+//
+// Copyright (c) 2015-2017, Deutsches Forschungszentrum für Künstliche Intelligenz GmbH.
+// Copyright (c) 2015-2017, University of Bremen
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <vizkit3d/TraversabilityGridVisualization.hpp>
+#include <vizkit3d/ExtentsRectangle.hpp>
+#include <vizkit3d/ColorConversionHelper.hpp>
+
+#include <osg/Texture2D>
+
+
+using namespace vizkit3d;
+
+TraversabilityGridVisualization::TraversabilityGridVisualization()
+    : showMapExtents(true),
+      interpolateCellColors(true),
+      useNPOTTextures(false),
+      unknownColor(osg::Vec4(0, 0, 1, 1)),
+      obstacleColor(osg::Vec4(0.8, 0, 0, 1)),
+      minTraversableColor(osg::Vec4(1, 0, 0, 1)),
+      maxTraversableColor(osg::Vec4(0, 1, 0, 1))
+{
+}
+
+TraversabilityGridVisualization::~TraversabilityGridVisualization()
+{
+}
+
+osg::ref_ptr<osg::Node> TraversabilityGridVisualization::createMainNode()
+{
+    // Geode is a common node used for vizkit3d plugins. It allows to display
+    // "arbitrary" geometries
+    osg::ref_ptr<osg::Group> group = new osg::Group();
+    osg::ref_ptr<osg::Geode> geode = new osg::Geode();
+    group->addChild(geode.get());
+
+    return group.release();
+}
+
+void TraversabilityGridVisualization::updateMainNode(osg::Node* node)
+{
+    osg::ref_ptr<osg::Group> group = node->asGroup();
+    group->removeChildren(0, group->getNumChildren());
```

Comment:
Ideally `createMainNode` creates as much of the osg graph as possible and `updateMainNode` only modifies existing objects and removes and creates new drawables when necessary.
Not that you can hold certain objects of the graph as members or your class if you need to identify them (With the constraint that they must only be modified when `updateMainNode` is called).